cdc/sinkv2/eventsink/txn/mysql/mysql.go

```diff
@@ -0,0 +1,728 @@
+	testDB, err := GetDBConnImpl(ctx, dsn.FormatDSN())
+	if err != nil {
+		return nil, err
+	}
+	defer testDB.Close()
+
+	// Adjust sql_mode for compatibility.
+	dsn.Params["sql_mode"], err = querySQLMode(ctx, testDB)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	dsn.Params["sql_mode"], err = dmutils.AdjustSQLModeCompatible(dsn.Params["sql_mode"])
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// NOTE: quote the string is necessary to avoid ambiguities.
+	dsn.Params["sql_mode"] = strconv.Quote(dsn.Params["sql_mode"])
+
+	dsnStr, err = generateDSNByParams(ctx, dsn, params, testDB)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// check if GBK charset is supported by downstream
+	gbkSupported, err := checkCharsetSupport(ctx, testDB, charset.CharsetGBK)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !gbkSupported {
+		log.Warn("gbk charset is not supported by downstream, "+
+			"some types of DDL may fail to be executed",
+			zap.String("hostname", hostName), zap.String("port", port))
+	}
+
+	db, err := GetDBConnImpl(ctx, dsnStr)
+	if err != nil {
+		return nil, err
+	}
+	db.SetMaxIdleConns(params.workerCount)
+	db.SetMaxOpenConns(params.workerCount)
+
+	ctx, cancel := context.WithCancel(ctx)
+	sink := &mysqlSink{
+		changefeedID:   changefeedID,
+		db:             db,
+		params:         params,
+		forceReplicate: replicaConfig.ForceReplicate,
+		enableOldValue: replicaConfig.EnableOldValue,
+		statistics:     metrics.NewStatistics(ctx, metrics.SinkTypeDB),
+		cancel:         cancel,
+	}
+
+	log.Info("mysql backend is created",
+		zap.String("changefeedID", fmt.Sprintf("%s.%s", changefeedID.Namespace, changefeedID.ID)),
+		zap.Bool("forceReplicate", sink.forceReplicate),
+		zap.Bool("enableOldValue", sink.enableOldValue))
+	return sink, nil
+}
+
+func (s *mysqlSink) OnTxnEvent(event *eventsink.TxnCallbackableEvent) (needFlush bool) {
+	if event.Event == nil {
+		return true
+	}
+	s.events = append(s.events, event)
+	s.rows += len(event.Event.Rows)
+	return event.Event.FinishWg != nil || s.rows >= s.params.maxTxnRow
+}
+
+func (s *mysqlSink) Flush(ctx context.Context) (err error) {
+	failpoint.Inject("SinkFlushDMLPanic", func() {
+		time.Sleep(time.Second)
+		log.Fatal("SinkFlushDMLPanic")
+	})
+	failpoint.Inject("MySQLSinkExecDMLError", func() {
+		// Add a delay to ensure the sink worker with `MySQLSinkHangLongTime`
+		// failpoint injected is executed first.
+		time.Sleep(time.Second * 2)
+		failpoint.Return(errors.Trace(dmysql.ErrInvalidConn))
+	})
+
+	for _, event := range s.events {
+		s.statistics.ObserveRows(event.Event.Rows...)
+	}
+
+	dmls := s.prepareDMLs()
+	log.Debug("prepare DMLs", zap.Any("rows", s.rows),
+		zap.Strings("sqls", dmls.sqls), zap.Any("values", dmls.values))
+
+	if err := s.execDMLWithMaxRetries(ctx, dmls); err != nil {
+		log.Error("execute DMLs failed", zap.String("err", err.Error()))
+		return errors.Trace(err)
+	}
+
+	s.events = s.events[:0]
```

Comment:
- Please zero the values of `s.events` before reslicing to length zero.
- Please check for `cap(s.events)`. If it's too large, we should set `s.events = nil` to allow the underlying array to be GC'ed.